### PR TITLE
Feature/contact history

### PR DESCRIPTION
**Pull request by pankaj-ag**
Issue: https://github.com/glific/glific/issues/1837

- First cut for capturing the contact history for various operations.
- We will add more data points in the future
- [x]  **Contact Created**
     event name: contact_created
     event label:  "contact created on 2 Dec, 2021"
- [x]  **Contact Field XX updated with value VV**
     event name: contact_fields_updated
     event label:  "Value for state is updated to Haryana"
- [x]  **Contact Language updated to LL**
     event name: contact_language_updated
     event label:  "contact language set to Hindi via flow Langauge Flow"
- [x]  **Contact Name updated to NN**
     event name: contact_name_updated
     event label:  "contact name changed to Ram"
- [x]  **Contact added/removed from Collection A**
     event name: contact_groups_updated
     event label:  "Contact added to the Collection  optin"
     event label:  "Contact removed to the group"
- [x]  **Contact started on Flow A**
     event name: contact_flow_started
     event label:  "Flow (Help workflow) started for the contact"
- [x]  **Contact ended Flow A**
     event name: contact_flow_ended
     event label:  "Flow (Help workflow) ended for the contact"
- [x]  **Contact ended all Flows**
     event name: contact_flow_ended_all
     event label:  "All flows are ended for the contact"

### CODE REVIEW

File: lib/glific/contacts.ex [hunk 1]
```diff
@@ -214,6 +215,13 @@ defmodule Glific.Contacts do
     %Contact{}
     |> Contact.changeset(attrs)
     |> Repo.insert()
+    |> case do
+      {:ok, contact} ->
+        {:ok, contact}
+
+      {:error, changeset} ->
+        {:error, changeset}
+    end
```
Comment (dlobo) [inline lib/glific/contacts.ex]: Why is this needed? since it basically returns the value of Repo.insert?

File: lib/glific/contacts.ex [hunk 2]
```diff
@@ -239,6 +247,13 @@ defmodule Glific.Contacts do
         contact
         |> Contact.changeset(attrs)
         |> Repo.update()
+        |> case do
```
Comment (dlobo) [inline lib/glific/contacts.ex]: Same question as above

File: lib/glific/contacts.ex [hunk 3]
```diff
@@ -691,4 +706,34 @@ defmodule Glific.Contacts do
   """
   @spec is_simulator_contact?(String.t()) :: boolean
   def is_simulator_contact?(phone), do: String.starts_with?(phone, @simulator_phone_prefix)
+
+  @doc """
+  create new contact histroy record.
+  """
+  @spec capture_history(Contact.t() | non_neg_integer(), atom(), map()) :: :ok
+  def capture_history(contact_id, event_type, attrs) when is_integer(contact_id) == true,
```
Comment (dlobo) [inline lib/glific/contacts.ex]: == true is redundant?

File: lib/glific/contacts.ex [hunk 4]
```diff
@@ -691,4 +706,36 @@ defmodule Glific.Contacts do
...
+          event_meta: %{}
+        },
+        attrs
+      )
+
+    %ContactHistory{}
+    |> ContactHistory.changeset(attrs)
+    |> Repo.insert()
+
+    :ok
```
Comment (dlobo) [inline lib/glific/contacts.ex]: why not return the value from Repo.insert. we should stop returning :ok and ignore insert results

File: lib/glific/flows/action.ex [hunk 5]
```diff
@@ -526,6 +536,16 @@ defmodule Glific.Flows.Action do
           action.groups,
           fn group ->
             {:ok, group_id} = Glific.parse_maybe_integer(group["uuid"])
+
+            Contacts.capture_history(context.contact_id, :contact_removed_from_group, %{
```
Comment (dlobo) [inline lib/glific/flows/action.ex]: ensure that {:ok, _} was returned so we can catch (and die) on the error

File: lib/glific/flows/contact_field.ex [hunk 6]
```diff
@@ -60,6 +71,10 @@ defmodule Glific.Flows.ContactField do
         %{fields: %{}}
       )
 
+    Contacts.capture_history(contact, :contact_field_reset, %{
+      event_name: "reset all conatct fields"
```
Comment (dlobo) [inline lib/glific/flows/contact_field.ex]: spelling mistake

File: lib/glific/flows/flow_context.ex [hunk 7]
```diff
@@ -429,6 +437,10 @@ defmodule Glific.Flows.FlowContext do
     |> where([fc], fc.is_background_flow == false)
     |> Repo.update_all(set: [completed_at: now, node_uuid: nil, updated_at: now])
 
+    Contacts.capture_history(contact_id, :contact_flow_stopped_all, %{
+      event_name: "all flow stopped for the contact"
```
Comment (dlobo) [inline lib/glific/flows/flow_context.ex]: we should decide the protocol and style of event name and standardize it

**Review by dlobo — CHANGES_REQUESTED** [review]

**Comment by mdshamoon** [comment]
@pankaj-ag feedbacks for contact history:
Link  for design file- [Design file](https://www.sketch.com/s/ba323b85-11d7-4687-8a19-681a32132bab/a/81A5a7)

1. Contact Added to collection -> Update to 'Added to collection: "Collection name in quotes" '
2. Contact Removed to collection -> Update to 'Removed from collection: "Collection name in quotes" '
3. Flow (flow name) started for the contact -> Flow Started  (I can get the flow name from event meta)


image

**Pull request merged**